enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -111,116 +128,113 @@ users' workload Pods so that containers in those Pods can utilize those CAs when
-
-**NO SecretProvideClass IS NEEDED**
-
-With those simplifying decisions, a CSI driver with those characteristics still meets our needs.  And the storage
-SMEs recommend that approach to us.
-
-Crafting the CSI driver in this way also helps us deal with SELinux.  Specifically (thanks to @jsafrane for details):
-
-- when CRI-O starts any container, it updates the SELinux labels on all volumes for that container, where those labels are unique to the container
-- in the context of k8s, multiple containers within a pod have the same filesystem access
-- so, only that specific pod can read files on that volume
-- however, a file can only have one SELinux label
-- so two or more pods cannot access the same file
-- but if each pod has its own file, its own copy, of the shared data, things work easily from a SELinux perspective
-- this is where using ephemeral CSI storage comes in
-- using ephemeral based storage allows for use of a tmpfs (virtual memory filesystem) for storage
-- tmpfs is fast and has negligible overhead
-- it will have to reconstructed if the Pod moves
-- but how it eases dealing with SELinux protection is the bonus we are taking
-- so each pod gets its own tmpfs
-- the kubelet leverages tmpfs in a similar fashion already
-
-#### End to end flow (from creation, to opt-in, to validation, and then injections)
+- Making it a day 1 install operation meets the ease of use targets we are trying to reach with the broader epics around
+Red Hat entitlements.
+- To paraphrase Clayton Coleman's comments during the decision-making process, this is the type of enhancement on top
+of Kubernetes that core OCP is all about.
+
+#### API, RBAC, Observability and Discoverability
 
 An admin creates a new cluster level custom resource for encapsulating the sharable `Secret` or `ConfigMap`.
 
 
 ```yaml
-kind: Share
-apiVersion: projected-resource.storage.openshift.io/v1
+kind: ShareResource
+apiVersion: storage.openshift.io/v1alpha1
 metadata:
   name: shared-cool-secret
 spec:
-  backingResource:
-    kind: Secret
-    apiVersion: v1
-    namespace: ns-one
-    name: cool-secret
-  description: "used to access my internal git repo, RHEL entitlements, or something else cool"
+  resource:
+    type: Secret
+    secret:
+      namespace: ns-one
+      name: cool-secret
 status:
   conditions:
   ...
 ```
 
-And sets up ACL/RBAC to control who can list/get such objects.
+And sets up ACL/RBAC to control which `ServiceAccounts` can use such objects to mount their referenced resources in their `Pods`:
 
 ```yaml
 rules:
 - apiGroups:
-  - projected-resource.storage.openshift.io
+  - storage.openshift.io
   resources:
-  - shares
+  - shareresources
   resourceNames:
   - shared-cool-secret
   verbs:
-  - get
+  - use
+
+```
+
+And for controlling which human users can list or inspect the `SharedResources` available on a cluster:
+
+```yaml
+rules:
+- apiGroups:
+  - storage.openshift.io
+  resources:
+  - shareresources
+  verbs:
   - list
-  - watch
+  - get
```

Comment:
I think there is an argument to be made that this cluster role should be installed by the operator and aggregated to the user-facing "edit" role [1]:

1. Edit users are assumed to create and manage workloads. By their very nature they are granted a high level of trust.
2. Edit users are the primary audience who need to be able to discover shared resources.
3. The details embedded in a SharedResource are IMO pseduo-sensitive, equivalent to a network IP address. The object can divulge the _potential_ existence of a Secret/ConfigMap in another namespace. However the object does not provide full details of truly sensitive information contained in a Secret.

I really do like this solution with the `use` verb - it succeeds in separating usage of the shared resource with the discovery of the shared resource. I can envision a `SharedResource` to be secured as follows:

1. Admins have full permissions (`*` verb)
2. Editors have (`get`, `list`), thereby can discover the `SharedResources`
3. Editors can request that a _service account_ be granted the `use` permission. Admin configures the appropriate cluster rolebinding.
4. Editors can then deploy a workload which _uses_ the shared resource. 

[1] https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles